rust/lance-encoding/src/encodings/logical/list.rs

```diff
@@ -416,8 +429,57 @@ struct IndirectlyLoaded {
     item_decoders: Vec<Box<dyn LogicalPageDecoder>>,
 }
 
+/// An encoder for list offsets that "stitches" offsets
+///
+/// If we need to encode several list arrays into a single page then we need to "stitch" the offsets
+/// For example, imagine we have list arrays [[0, 1], [2]] and [[3, 4, 5]].
+///
+/// We will have offset arrays [0, 2, 3] and [0, 3].  We don't want to encode [0, 2, 3, 0, 3].  What
+/// we want is [0, 2, 3, 6]
+#[derive(Debug)]
+struct ListOffsetsEncoder {
+    inner: Box<dyn ArrayEncoder>,
+}
+
+impl ArrayEncoder for ListOffsetsEncoder {
+    fn encode(&self, arrays: &[ArrayRef], buffer_index: &mut u32) -> Result<EncodedArray> {
+        if arrays.len() < 2 {
+            // Nothing to patch, don't incur a copy
+            return self.inner.encode(arrays, buffer_index);
+        }
+        println!("Stitching offsets {:?}", arrays);
+        let num_offsets =
+            arrays.iter().map(|array| array.len()).sum::<usize>() - (arrays.len() - 1);
+        let mut offsets = Vec::with_capacity(num_offsets);
+        offsets.extend_from_slice(arrays[0].as_primitive::<Int32Type>().values());
+        for array in &arrays[1..] {
+            let last_prev_offset = *offsets.last().unwrap();
+            let values = array.as_primitive::<Int32Type>().values();
+            if values.len() == 0 {
+                continue;
+            }
+            // The first offset doesn't have to be 0 (this happens when lists are sliced)
+            //
+            // So if the previous offsets are [0, 3, 5] and the current offsets are
+            // [10, 11, 12] then we want to add 5 - 10 to all of the current offsets
+            // (skipping the first) to get [6, 7] and then append them to the previous
+            // offsets to get [0, 3, 5, 6, 7]
```

Comment:
Are we handling the case where the first set of offsets doesn't start at 0?

Reply:
Yes, and the slicing tests in random will test this.